go/vt/vtgate/vindexes/cfc.go

```diff
@@ -0,0 +1,326 @@
+// Verify returns true if ids maps to ksids.
+func (vind *CFC) Verify(_ VCursor, ids []sqltypes.Value, ksids [][]byte) ([]bool, error) {
+	out := make([]bool, len(ids))
+	for i := range ids {
+		v, err := vind.computeKsid(ids[i].ToBytes(), false)
+		if err != nil {
+			return nil, err
+		}
+		out[i] = bytes.Equal(v, ksids[i])
+	}
+	return out, nil
+}
+
+// Map can map ids to key.Destination objects.
+func (vind *CFC) Map(cursor VCursor, ids []sqltypes.Value) ([]key.Destination, error) {
+	out := make([]key.Destination, len(ids))
+	for i, id := range ids {
+		v, err := vind.computeKsid(id.ToBytes(), false)
+		if err != nil {
+			return nil, err
+		}
+		out[i] = key.DestinationKeyspaceID(v)
+	}
+	return out, nil
+}
+
+// PrefixVindex switches the vindex to prefix mode
+func (vind *CFC) PrefixVindex() SingleColumn {
+	return vind.prefixVindex
+}
+
+func addOne(value []byte) []byte {
+	n := len(value)
+	overflow := true
+	for i := n - 1; i >= 0; i-- {
+		if value[i] < 255 {
+			value[i]++
+			overflow = false
+			break
+		} else {
+			value[i] = 0
+		}
+	}
+	if overflow {
+		return nil
+	}
+	return value
+}
+
+type prefixCFC struct {
+	name string
+	cfc  *CFC
+}
+
+func (vind *prefixCFC) String() string {
+	return vind.name
+}
+
+func (vind *prefixCFC) Cost() int {
+	return 1
+}
+
+func (vind *prefixCFC) IsUnique() bool {
+	return false
+}
+
+func (vind *prefixCFC) NeedsVCursor() bool {
+	return false
+}
+
+// Verify returns true if ids maps to ksids.
+func (vind *prefixCFC) Verify(vc VCursor, ids []sqltypes.Value, ksids [][]byte) ([]bool, error) {
+	// prefixCFC is only active for 'LIKE' expr. Verify is used by 'INSERT' so we re-use CFC.Verify()
+	return vind.cfc.Verify(vc, ids, ksids)
+}
+
+// Map can map ids to key.Destination objects.
+func (vind *prefixCFC) Map(cursor VCursor, ids []sqltypes.Value) ([]key.Destination, error) {
+	out := make([]key.Destination, len(ids))
+	for i, id := range ids {
+		value := id.ToBytes()
+		// we only consider the string up to the first whild char as the prefix.
+		p := bytes.IndexByte(value, '%')
+		if p < 0 {
+			// if there is no wild char, we use the whole string
+			p = len(value)
+		}
+		begin, err := vind.cfc.computeKsid(value[:p], true)
+		if err != nil {
+			return nil, err
+		} else if len(begin) == 0 {
+			// prefix is too short, this is typically the case when the prefix
+			// is smaller than the first offset when using hash.
+			out[i] = key.DestinationAllShards{}
+		} else {
+			// the prefix maps to keyspace id corresponding to value and plus one.
+			// that is [ keyspace_id, keyspace_id + 1 ).
+			end := make([]byte, len(begin))
+			copy(end, begin)
+			end = addOne(end)
+			out[i] = key.DestinationKeyRange{
+				KeyRange: &topodatapb.KeyRange{
+					Start: begin,
+					End:   end,
+				},
+			}
+		}
+
+	}
+	return out, nil
+}
+
+// we don't use the full hashed value because it's very long.
+// keyrange resolution is done via comparing []byte so longer
+// keyspace ids have performance impact.
```

Comment:
If you use hashing functions that are already in use for vindexes, then that shouldn't be an issue, right? If they were already too long, then they would already be causing issues.

Maybe smaller, more performant hash functions are needed just for cases where you are concatenating more than 3 values together, but for the common use-cases of 2 or 3 values, maybe an ordinary vindex hash function, in its entirety, would do?

Reply:
md5 hash is 16-byte. If there are only 2 or 3 components in the composite index, you are right that it doesn't matter much. However, the interface here allows arbitrarily concatenated composite index. Hence this comment is about this generic case. Specifically it makes the hashed value (of the concatenation) at most same size as the input value so that there is no added performance cost in `bytes.Compare` if people choose to use many components in the composite index.